presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java

```diff
@@ -0,0 +1,412 @@
+                        .putAll(toAssignments(aggregationNode.get().getOutputSymbols()))
+                        .putAll(subqueryProjection.get().getAssignments())
+                        .build();
+
+                return new ProjectNode(
+                        idAllocator.getNextId(),
+                        aggregationNode.get(),
+                        assignments);
+            }
+            else {
+                return aggregationNode.get();
+            }
+        }
+
+        private Optional<AggregationNode> createAggregationNode(
+                AggregationNode scalarAggregation,
+                JoinNode leftOuterJoin,
+                Symbol nonNullableAggregationSourceSymbol)
+        {
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregations = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Signature> functions = ImmutableMap.builder();
+            FunctionRegistry functionRegistry = metadata.getFunctionRegistry();
+            for (Map.Entry<Symbol, FunctionCall> entry : scalarAggregation.getAggregations().entrySet()) {
+                FunctionCall call = entry.getValue();
+                QualifiedName count = QualifiedName.of("count");
+                Symbol symbol = entry.getKey();
+                if (call.getName().equals(count)) {
+                    aggregations.put(symbol, new FunctionCall(
+                            count,
+                            ImmutableList.of(nonNullableAggregationSourceSymbol.toSymbolReference())));
+                    List<TypeSignature> scalarAggregationSourceTypeSignatures = ImmutableList.of(
+                            symbolAllocator.getTypes().get(nonNullableAggregationSourceSymbol).getTypeSignature());
+                    functions.put(symbol, functionRegistry.resolveFunction(
+                            count,
+                            scalarAggregationSourceTypeSignatures,
+                            false));
+                }
+                else {
+                    aggregations.put(symbol, entry.getValue());
+                    functions.put(symbol, scalarAggregation.getFunctions().get(symbol));
+                }
+            }
+
+            List<Symbol> groupBySymbols = leftOuterJoin.getLeft().getOutputSymbols();
+            return Optional.of(new AggregationNode(
+                    idAllocator.getNextId(),
+                    leftOuterJoin,
+                    groupBySymbols,
+                    aggregations.build(),
+                    functions.build(),
+                    scalarAggregation.getMasks(),
+                    ImmutableList.of(groupBySymbols),
+                    scalarAggregation.getStep(),
+                    scalarAggregation.getSampleWeight(),
+                    scalarAggregation.getConfidence(),
+                    scalarAggregation.getHashSymbol()));
+        }
+    }
+
+    private static class ScalarAggregationSourceRewriter
```

Comment:
I don't think we need a class for this. There's no state that needs to be maintained, so it could just be a bunch of static methods in the enclosing class.